src/hotspot/share/opto/mulnode.cpp

```diff
@@ -2051,99 +2053,75 @@ const Type* RotateRightNode::Value(PhaseGVN* phase) const {
-  if (phase->type(shift)->isa_integer(bt) == nullptr) {
-    return false;
+
+  if (type->is_con()) {
+    jlong con = type->get_con_as_long(bt);
+    return con == 0L ? (type2aelembytes(bt) * BitsPerByte) : count_trailing_zeros(con);
   }
-  BasicType shift_bt = bt;
-  if (bt == T_LONG && shift->Opcode() == Op_ConvI2L) {
+
+  if (expr->Opcode() == Op_ConvI2L) {
+    expr = expr->in(1)->uncast();
     bt = T_INT;
-    Node* val = shift->in(1);
-    if (val == nullptr) {
-      return false;
-    }
-    val = val->uncast();
-    if (val == nullptr) {
-      return false;
-    }
-    if (val->Opcode() == Op_LShiftI) {
-      shift_bt = T_INT;
-      shift = val;
-      if (phase->type(shift)->isa_integer(bt) == nullptr) {
-        return false;
-      }
-    }
+    type = phase->type(expr)->isa_int();
   }
-  if (shift->Opcode() != Op_LShift(shift_bt)) {
-    if (check_reverse &&
-        (mask->Opcode() == Op_LShift(bt) ||
-         (bt == T_LONG && mask->Opcode() == Op_ConvI2L))) {
-      // try it the other way around
-      return AndIL_shift_and_mask_is_always_zero(phase, mask, shift, bt, false);
+
+  if (expr->Opcode() == Op_LShift(bt)) {
+    const TypeInt* rhs_t = phase->type(expr->in(2))->isa_int();
+    if (rhs_t == nullptr || !rhs_t->is_con()) {
+      return -1;
     }
-    return false;
-  }
-  Node* shift2 = shift->in(2);
-  if (shift2 == nullptr) {
-    return false;
+    return rhs_t->get_con() & (type2aelembytes(bt) * BitsPerByte - 1);
   }
-  const Type* shift2_t = phase->type(shift2);
-  if (!shift2_t->isa_int() || !shift2_t->is_int()->is_con()) {
+
+  return -1;
+}
+
+// Checks whether expr is neutral wrt addition under mask, i.e.
+// an expression of the form (AndX (T+expr) mask) can be simplified to (AndX T mask).
+// The X in AndX must be I or L, depending on bt.
+// Specifically, this holds for the following cases,
+// when the shift value N is large enough to zero out
+// all the set positions of the and-mask M:
+//   (AndI (LShiftI _ #N) #M)
+//   (AndL (LShiftL _ #N) #M)
+//   (AndL (ConvI2L (LShiftI _ #N)) #M)
+// including equivalent constant operands:
+//   (AndI (ConI (_ << #N)) #M)
+//   (AndL (ConL (_ << #N)) #M)
+// The M and N values must satisfy ((-1 << N) & M) == 0.
+static bool AndIL_is_zero_element_under_mask(const PhaseGVN* phase, const Node* expr, const Node* mask, BasicType bt) {
+  jint expr_trailing_zeros = AndIL_min_trailing_zeros(phase, expr, bt);
+  if (expr_trailing_zeros < 0) {
     return false;
   }
 
-  jint shift_con = shift2_t->is_int()->get_con() & ((shift_bt == T_INT ? BitsPerJavaInteger : BitsPerJavaLong) - 1);
-  if ((((jlong)1) << shift_con) > mask_t->hi_as_long() && mask_t->lo_as_long() >= 0) {
-    return true;
+  const TypeInteger* mask_t = phase->type(mask)->isa_integer(bt);
+  if (mask_t == nullptr || mask_t->lo_as_long() < 0) {
+    return false;
   }
 
-  return false;
+  jint mask_bit_width = mask_t->hi_as_long() == 0 ? 0 : (BitsPerLong - count_leading_zeros(mask_t->hi_as_long()));
```

Comment:
I would just split this into 2 separate things.
```suggestion
  // Is the mask always zero?
  if (mask_t->hi_as_long() == 0) {
    assert(mask_t->lo_as_long() == 0, "checked earlier");
    return true;
  }
  jint mask_bit_width =  BitsPerLong - count_leading_zeros(mask_t->hi_as_long());
```